### PR TITLE
feat: added treesitter source

### PR DESCRIPTION
**Pull request by elianiva**
I don't use it personally but if people coming from completion.nvim want to have this source on nvim-compe as well, might worth to merge this in. Otherwise, feel free to close this PR if no one wants it.

### CODE REVIEW

**Comment by weilbith** [comment]
Sorry, this is not really related to this PR, but it popped up in my notifications and you might be able to give some insights.
What I'm currently missing and `completion.nvim` is the feature of independent asynchronous sources. So I have some big projects where the language server slows down. What I would like to have is tree-sitter as a fast completion source with low priority level and native LSP with high priority. So in case that LSP is slow, I can still get fast on-the-fly completion suggestions (e.g. long variable names) via Treesitter. But when LSP resolves its candidate, they get added to the list and take higher priority in the sorting order than the Treesitter once as I argue they are more accurate. An example would be the trying of a class instance variable and then waiting for the completion of possible methods of the class or its properties etc. Just because you don't remember the library anymore or what ever.
Is this kinda possible with this completion plugin?

**Comment by elianiva** [comment]
You can already do that, I think. Just make sure treesitter source has lower priority than LSP source (which I already did if I remember correctly, so no configuration needed)

**Comment by cloggier** [comment]
It would be nice if it then removed lower priority items if they are duplicates. Just a thought.

**Comment by hrsh7th** [comment]
The all of above discussion is the same to current behavior.

**Comment by weilbith** [comment]
Wow, cool this sounds awesome.
So just completion sources are treated independent in this plugin and the completion menu gets populated when the first source responds and then gets continuously extended each time another source responds?

**Comment by weilbith** [comment]
> The all of above discussion is the same to current behavior.

As I stated this was not directly targeting this PR but being a question about the plugin.

**Comment by hrsh7th** [comment]
Almost yes, But sometimes it makes flicker so this plugin has `source_timeout` option.

**Comment by weilbith** [comment]
Cool. If this is possible you got me and I'll switch other to this plugin soon. This is what I was awaiting for. :hugs: :partying_face:

**Comment by hrsh7th** [comment]
I will review it after experiencing it.

**Comment by hrsh7th** [comment]
I've tried it. It is useful than my thoughts.

File: after/plugin/compe_treesitter.vim [hunk 1]
```diff
@@ -0,0 +1,9 @@
+if exists('g:loaded_compe_treesitter')
+  finish
+endif
+let g:loaded_compe_treesitter = v:true
+
+if exists('g:loaded_compe')
```
Comment (hrsh7th) [inline after/plugin/compe_treesitter.vim:6]: we should check `nvim-treesitter` is installed or not.
(You can check it in here or compe_treesitter/init.lua)

File: lua/compe_treesitter/init.lua [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      for _, match in ipairs(matches) do
+        local node = match.node
+        local node_scope = self:_get_smallest_context(node)
+        local start_line_node, _, _= node:start()
+        local node_text = ts_utils.get_node_text(node, 0)[1]
+
+        if node_text
+          and (not node_scope or ts_utils.is_parent(node_scope, at_point))
+          and (start_line_node <= line_current) then
+          table.insert(complete_items, {word = node_text, kind = match.kind})
```
Comment (hrsh7th) [inline lua/compe_treesitter/init.lua:68]: The kind is number so it is not useful... Hm... we should define the mapping?

File: lua/compe_treesitter/init.lua [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    for name, item in pairs(match) do
+      vim.list_extend(matches, self:_prepare_match(item, name))
+    end
+  end
+
+  return matches
+end
+
+function Source._get_treesitter_items(self)
+  if parsers.has_parser() then
```
Comment (hrsh7th) [inline lua/compe_treesitter/init.lua:50]: This check should move to `determine`.

File: lua/compe_treesitter/init.lua [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    return complete_items
+  end
+
+  return {}
+end
+
+function Source.complete(self, context)
+  context.callback({
+    items = self:_get_treesitter_items(),
+    incomplete = true
```
Comment (hrsh7th) [inline lua/compe_treesitter/init.lua:82]: I think it is not necessary.

File: lua/compe_treesitter/init.lua (hunk 2 above)
Comment (elianiva) [reply lua/compe_treesitter/init.lua:68]: I agree, it is not *that* useful at the moment

**Comment by elianiva** [comment]
Thanks for the review! I'll fix them tomorrow since I won't be able to access my laptop today.

File: lua/compe_treesitter/init.lua (hunk 2 above)
Comment (elianiva) [reply lua/compe_treesitter/init.lua]: do  you still see the kind field filled with number? I see them as `var`, `import`, `associated`, etc.

Comment (hrsh7th) [reply lua/compe_treesitter/init.lua]: I can't see the number at the moment. Thank you.

File: lua/compe_treesitter/init.lua [hunk 5]
```diff
@@ -0,0 +1,87 @@
+local ts_exists = pcall(require, "nvim-treesitter")
```
Comment (hrsh7th) [inline lua/compe_treesitter/init.lua:1]: I think this is not enough because the still require `nvim-treesitter.locals` without pcall.

Comment (elianiva) [reply lua/compe_treesitter/init.lua:1]: I think if someone have nvim-treesitter installed, they will always have `nvim-treesitter.locals` so I assumed we only need to check the top module.

Comment (hrsh7th) [reply lua/compe_treesitter/init.lua:1]: If the user does not have `nvim-treesitter`, `require'nvim-treesitter.locals` will throw exception.

1. `local ts_exists = pcall(require, "nvim-treesitter")` ignore exception.
2. `local ts_locals = require('nvim-treesitter.locals')`

Currently, nvim-compe uses `pcall` for public apis.

Could you check with `let g:compe.debug = v:true`? (If you have a time)

Comment (elianiva) [reply lua/compe_treesitter/init.lua:1]: ah, yeah, that makes sense. Will fix it in a moment

Comment (hrsh7th) [reply lua/compe_treesitter/init.lua:1]: Good! Thank you!

**Comment by hrsh7th** [comment]
It almost looks good to me.

I'm merging it. if someone wants to improve it more, please create PR.

**Comment by hrsh7th** [comment]
@elianiva Thank you very much! I'm happy for your contribution.

**Pull request merged**